### PR TITLE
GROK-13392: Hit triage: Add alerts, toxicity, descriptors and properties calculation

### PR DESCRIPTION
**Pull request by drizhina**

### CODE REVIEW

File: packages/Chem/src/OCL-service/OCL-service.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
+import * as DG from 'datagrok-api/dg';
+import {OCLServiceCall} from './consts';
+
+export class OCLService {
+  _workers: Worker[];
+  _threadCount: number = 4;
```
Comment (skalkin) [inline packages/Chem/src/OCL-service/OCL-service.ts:6]: Why not initialize this field fully right here?

File: packages/Chem/src/OCL-service/OCL-service.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  constructor() {
+    this._threadCount = Math.max((navigator.hardwareConcurrency || 6) - 2, 1);
+    this._workers = new Array(this._threadCount).fill(null)
+      .map(() => new Worker(new URL('OCL-worker', import.meta.url)));
+  }
+
+  private async _doParallel(
+    molCol:DG.Column, argList: any[], op: OCLServiceCall,
+  ): Promise<{[key: string]: Array<number>}> {
+    const molList = molCol.toList();
```
Comment (skalkin) [inline packages/Chem/src/OCL-service/OCL-service.ts:16]: Do we really need to create another array? Is it just for being able to use the `slice` method?

File: packages/Chem/src/OCL-service/OCL-service.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        this._workers[i].onmessage = ({data: {res, errors}}) => {
+          resolve({res, errors});
+        };
+      });
+    }
+    const scatterRes = await Promise.all(promises);
+    const errorPercentages = scatterRes.map(({errors}) => errors.length)
+      .reduce((prev, cur) => prev + cur, 0) / colLenght;
+    if (errorPercentages > 0.1)
+      throw new Error('Too many errors occured during calculation');
```
Comment (skalkin) [inline packages/Chem/src/OCL-service/OCL-service.ts:34]: It has to be handled better (operation succeeding and # of errors shown)

File: packages/Chem/src/OCL-service/OCL-worker.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    postMessage(getChemProperties(data, argList));
+    break;
+  default:
+    postMessage(getToxRisks(data, argList));
+    break;
+  }
+};
+
+
+function getChemProperties(molList: Array<string>, propList: string[]) {
```
Comment (skalkin) [inline packages/Chem/src/OCL-service/OCL-worker.ts:16]: Please annotate return type

File: packages/Chem/src/OCL-service/OCL-worker.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  const res: {[key: string]: Array<number>} = {};
+  const errors: string[] = [];
+  propList.forEach((propName) => {
```
Comment (skalkin) [inline packages/Chem/src/OCL-service/OCL-worker.ts:19]: Why not use the standard for method, it's shorter and a bit more expressive?

```
for (let propName of propList)
  res[propName] = new Array(molList.length).fill(NaN);
```

File: packages/Chem/src/OCL-service/OCL-worker.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        res[p][i] = CHEM_PROP_MAP[p].valueFunc(mol);
+      });
+    } catch (e) {
+      errors.push(e instanceof Error ? e.message : e as string);
+    }
+  });
+  return {res, errors};
+}
+
+function getToxRisks(molList: Array<string>, riskTypes: number[]) {
```
Comment (skalkin) [inline packages/Chem/src/OCL-service/OCL-worker.ts:35]: Please provide return type

File: packages/Chem/src/widgets/properties.ts [hunk 7]
```diff
@@ -95,11 +70,24 @@ export function propertiesWidget(semValue: DG.SemanticValue<string>): DG.Widget
...
+
+export async function addPropertiesAsColumns(df: DG.DataFrame, smilesCol: DG.Column<string>,
+  propsMap: string[]) {
+  const oclService = new OCLService();
+  const props = await oclService.getChemProperties(smilesCol, propsMap);
+
+  oclService.terminate();
+  propsMap.forEach((p) => {
+    const colName = df.columns.getUnusedName(p);
+    const col = DG.Column.fromList(DG.TYPE.FLOAT, colName, props[p]);
```
Comment (skalkin) [inline packages/Chem/src/widgets/properties.ts:90]: Some of the properties are ints

File: packages/HitTriage/src/views/2-compute-view.ts [hunk 8]
```diff
@@ -1,26 +1,51 @@
...
+import {chemDescriptorsDialog} from './descriptors-dialog';
+import {ChemPropNames} from './types';
+import {ChemPropertyGroupMap} from './consts';
 
 /**
  * Enrichment of the molecular dataset.
  **/
-export class ComputeView extends HitTriageBaseView {
 
+const chemFunctionsMap: {[_ in ChemPropNames]: (t: DG.DataFrame, col: string, props: string[]) => Promise<any>} = {
```
Comment (skalkin) [inline packages/HitTriage/src/views/2-compute-view.ts:14]: Would be nice to assembly this dynamically using Func.find

File: packages/HitTriage/src/views/2-compute-view.ts [hunk 9]
```diff
@@ -1,26 +1,51 @@
...
+  'Descriptors': (t, c, p) => grok.chem.descriptors(t, c, p),
+  'Toxicity Risks': (t, c, p) => grok.functions.call(
+    'Chem:addChemRisksColumns', getChemFunctionArgs(t, c, p, 'Toxicity Risks')),
+  'Structural Alerts': (t, c, p) => grok.functions.call(
+    'Chem:structuralAlertsTopMenu', getChemFunctionArgs(t, c, p, 'Structural Alerts')),
+  'Chemical Properties': (t, c, p) => grok.functions.call(
+    'Chem:addChemPropertiesColumns', getChemFunctionArgs(t, c, p, 'Chemical Properties')),
+};
+
+function getChemFunctionArgs(t: DG.DataFrame, col: string, props: string[], funcName: ChemPropNames) {
```
Comment (skalkin) [inline packages/HitTriage/src/views/2-compute-view.ts:24]: Please annotate return type

**Review by skalkin — CHANGES_REQUESTED** [review]
Looks good to me! Let's address my comments, and then merge with master.

File: packages/Chem/src/OCL-service/OCL-service.ts (hunk 2 above)
Comment (drizhina) [reply packages/Chem/src/OCL-service/OCL-service.ts:16]: yes

**Pull request merged**